From 6127c206af0c0a4734c0e8725d5e0be051b22353 Mon Sep 17 00:00:00 2001
From: Sriram R <quic_srirrama@quicinc.com>
Date: Tue, 26 Sep 2023 18:44:06 +0530
Subject: [PATCH] wifi: ath12k: Fix beacon rx for sta associated to non Tx AP

Currently, When a station is associated to a non Tx AP, beacons are not
received frequently from FW. This results in missing events via beacons
such as channel switch leading to sta vif not switching to new channel as
the AP and causing kickout event from FW eventually.

This was due to missing configuration of non Tx info during STA vdev up.
Fill the info such as Tx bssid, profile index and count for addressing this
issue.

Signed-off-by: Sriram R <quic_srirrama@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 0435af1..2757e1b 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -4026,6 +4026,13 @@ static void ath12k_bss_assoc(struct ath12k *ar,
 	params.vdev_id = arvif->vdev_id;
 	params.aid = ahvif->aid;
 	params.bssid = arvif->bssid;
+
+	if (bss_conf->nontransmitted) {
+		params.profile_idx = bss_conf->bssid_index;
+		params.profile_count = BIT(bss_conf->bssid_indicator) - 1;
+		params.tx_bssid = bss_conf->transmitter_bssid;
+	}
+
 	ret = ath12k_wmi_vdev_up(ar, &params);
 	if (ret) {
 		ath12k_warn(ar->ab, "failed to set vdev %d up: %d\n",
@@ -11716,6 +11723,13 @@ beacon_tmpl_setup:
 		params.profile_idx = ahvif->vif->bss_conf.bssid_index;
 		params.profile_count = tx_arvif->nontransmitting_vif_count;
 	}
+
+	if (ahvif->vif->type == NL80211_IFTYPE_STATION && link->nontransmitted) {
+		params.profile_idx = link->bssid_index;
+		params.profile_count = BIT(link->bssid_indicator) - 1;
+		params.tx_bssid = link->transmitter_bssid;
+	}
+
 	rcu_read_unlock();
 	ret = ath12k_wmi_vdev_up(arvif->ar, &params);
 	if (ret) {
-- 
2.17.1

